TUS IMPLEMENTATION

zerox80 opened the pull request:
# Implementierung von wiederaufnehmbaren Uploads mit dem TUS-Protokoll

## Zusammenfassung

Dieses Pull-Request führt die Unterstützung für das TUS-Protokoll (v1.0.0) für Datei-Uploads ein. Dies ermöglicht robuste, wiederaufnehmbare und in Blöcke aufgeteilte Uploads, was die Zuverlässigkeit bei großen Dateien und instabilen Netzwerkverbindungen erheblich verbessert. Die Implementierung ist in das bestehende [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager)-System für Hintergrund-Uploads integriert und beinhaltet einen Fallback-Mechanismus auf traditionelle WebDAV-Uploads, falls TUS nicht verfügbar ist.

## Problembeschreibung

Das Hochladen großer Dateien auf mobilen Geräten ist anfällig für Fehler durch Netzwerkunterbrechungen, das Schließen der App oder Server-Timeouts. Der bisherige, auf WebDAV basierende Upload-Mechanismus unterstützte keine Wiederaufnahme von Uploads. Das bedeutete, dass jede Unterbrechung einen vollständigen Neustart des Uploads erforderte. Dies führte zu einer schlechten Benutzererfahrung, verschwendeter Bandbreite und einer hohen Fehlerquote bei großen Dateien.

## Lösungsansatz

Dieses PR löst das Problem, indem das TUS-Protokoll als primäre Methode für Uploads implementiert wird, die eine bestimmte Chunk-Größe überschreiten.

### Wesentliche Änderungen im Detail:

*   **TUS-Protokoll-Integration**:
    *   Neue `RemoteOperation`-Klassen für die zentralen TUS-Aktionen wurden hinzugefügt:
        *   [opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/CreateTusUploadRemoteOperation.kt](cci:7://file:///c:/Users/Rujbin.Nassereslam/Downloads/android-main%20%2812%29/android-main/opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/CreateTusUploadRemoteOperation.kt:0:0-0:0)
        *   [opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/PatchTusUploadChunkRemoteOperation.kt](cci:7://file:///c:/Users/Rujbin.Nassereslam/Downloads/android-main%20%2812%29/android-main/opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/PatchTusUploadChunkRemoteOperation.kt:0:0-0:0)
        *   [opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/GetTusUploadOffsetRemoteOperation.kt](cci:7://file:///c:/Users/Rujbin.Nassereslam/Downloads/android-main%20%2812%29/android-main/opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/GetTusUploadOffsetRemoteOperation.kt:0:0-0:0)
        *   [opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/CheckTusSupportRemoteOperation.kt](cci:7://file:///c:/Users/Rujbin.Nassereslam/Downloads/android-main%20%2812%29/android-main/opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/CheckTusSupportRemoteOperation.kt:0:0-0:0)
    *   TUS-spezifische HTTP-Header (`Tus-Resumable`, `Upload-Offset`, etc.) wurden in [opencloudComLibrary/src/main/java/eu/opencloud/android/lib/common/http/HttpConstants.java](cci:7://file:///c:/Users/Rujbin.Nassereslam/Downloads/android-main%20%2812%29/android-main/opencloudComLibrary/src/main/java/eu/opencloud/android/lib/common/http/HttpConstants.java:0:0-0:0) definiert.

*   **Anpassungen der Worker-Klassen**:
    *   Die zentrale Upload-Logik in [opencloudApp/src/main/java/eu/opencloud/android/workers/UploadFileFromFileSystemWorker.kt](cci:7://file:///c:/Users/Rujbin.Nassereslam/Downloads/android-main%20%2812%29/android-main/opencloudApp/src/main/java/eu/opencloud/android/workers/UploadFileFromFileSystemWorker.kt:0:0-0:0) und [opencloudApp/src/main/java/eu/opencloud/android/workers/UploadFileFromContentUriWorker.kt](cci:7://file:///c:/Users/Rujbin.Nassereslam/Downloads/android-main%20%2812%29/android-main/opencloudApp/src/main/java/eu/opencloud/android/workers/UploadFileFromContentUriWorker.kt:0:0-0:0) wurde überarbeitet, um TUS für Dateien zu priorisieren, die größer als `CHUNK_SIZE` sind.
    *   **Wiederaufnahmefähigkeit**: Die TUS-Upload-URL und der aktuelle Fortschritt (Offset) werden nun im `opencloudDomain/src/main/java/eu/opencloud/android/domain/transfers/TransferRepository.kt` in der Datenbank gespeichert. Bei einer Unterbrechung kann der Worker den Upload genau an der letzten bekannten Position fortsetzen.
    *   **Wiederholungslogik mit Backoff**: Die `PATCH`-Operation zum Hochladen der Chunks enthält nun einen robusten Wiederholungsmechanismus. Bei vorübergehenden Netzwerkfehlern wird der Versuch nach einer exponentiell ansteigenden Wartezeit wiederholt, um die Stabilität zu erhöhen.
    *   **Fallback-Mechanismus**: Falls der Server TUS nicht unterstützt oder die initiale Erstellung des TUS-Uploads fehlschlägt, schalten die Worker automatisch auf den bisherigen WebDAV-basierten Chunk-Upload um. Dies stellt die Abwärtskompatibilität sicher.

*   **Datenbank-Erweiterungen**:
    *   Das `opencloudDomain/src/main/java/eu/opencloud/android/domain/transfers/model/OCTransfer.kt`-Datenmodell und das `TransferRepository` wurden um Felder zur Speicherung des TUS-Zustands erweitert (`tusUploadUrl`, `tusUploadOffset`, etc.), um die Wiederaufnahmefähigkeit über App-Neustarts hinweg zu gewährleisten.

*   **UI- und Fortschrittsanzeige**:
    *   Der Upload-Fortschritt wird über die `WorkManager`-API direkt aus der TUS-Upload-Schleife an die Benutzeroberfläche ([opencloudApp/src/main/java/eu/opencloud/android/ui/activity/UploadListActivity.java](cci:7://file:///c:/Users/Rujbin.Nassereslam/Downloads/android-main%20%2812%29/android-main/opencloudApp/src/main/java/eu/opencloud/android/ui/activity/UploadListActivity.java:0:0-0:0)) gemeldet. Dies sorgt für eine nahtlose und korrekte Fortschrittsanzeige.
    *   Die Fehlerbehandlung wurde verbessert, um spezifischeres Feedback bei Upload-Fehlern (z.B. bei Anmeldefehlern) zu geben.

*   **Logging**:
    *   Der gesamte TUS-Workflow wurde mit ausführlichem Debug-Logging via [`Timber`](https://github.com/JakeWharton/timber) versehen, um die Fehlersuche und Überprüfung zu vereinfachen.

### Anleitung zum Testen

1.  **Erfolgreicher TUS-Upload**:
    *   Eine Datei hochladen, die größer als `CHUNK_SIZE` ist.
    *   Überprüfen, ob der Upload erfolgreich abgeschlossen wird.
    *   In den Logs nach Meldungen suchen, die auf die Verwendung von TUS hinweisen (z.B. "Attempting TUS for large upload", "TUS resume offset").

2.  **Test der Wiederaufnahmefähigkeit**:
    *   Den Upload einer großen Datei starten.
    *   Während des Uploads die Netzwerkverbindung trennen.
    *   Nach kurzer Zeit die Verbindung wiederherstellen.
    *   Beobachten, ob der Upload automatisch an der unterbrochenen Stelle fortgesetzt und erfolgreich beendet wird.

3.  **Test des Fallback-Mechanismus**:
    *   (Falls möglich) Gegen einen Server testen, auf dem TUS deaktiviert ist.
    *   Eine große Datei hochladen.
    *   Überprüfen, ob der Upload erfolgreich über den alten Chunk-Mechanismus abgeschlossen wird. In den Logs sollte eine Meldung wie "TUS flow failed, will fallback..." erscheinen.

4.  **Test bei App-Neustart**:
    *   Einen großen Upload starten.
    *   Die App vollständig schließen (Force Close).
    *   Die App neu öffnen.
    *   Überprüfen, ob der Upload automatisch fortgesetzt wird.

zerox80 commented:
opencloudApp/src/main/java/eu/opencloud/android/datamodel/ThumbnailsCacheManager.java:72 ergänzt einen threadsicheren In-Memory-LruCache samt Lock, prüft Schlüssel/Bitmap vor dem Schreiben, generiert robuste Cache-Keys für OCFile und fällt beim Disk-Cache auf das interne Verzeichnis zurück, falls extern nicht verfügbar.
In derselben Klasse (181, 408-465, 523-608) wird die Thumbnail-Pipeline gehärtet: Streams werden in finally geschlossen, HTTP-404 führt sofort zum Abschalten weiterer Aktualisierungen, PNGs erhalten den Hintergrundeinzug, und executeThumbnailTask bündelt AsyncTask-Aufrufe auf den Thread-Pool.
opencloudApp/src/main/java/eu/opencloud/android/presentation/files/filelist/FileListAdapter.kt:68-191 speichert das Ergebnis von shouldDisallowTouches…, aktiviert stabile IDs (getItemId), blendet Footer in Picker-Szenarien aus und korrigiert Selektionshilfen (selectableItemCount, versteckte Drei-Punkt-Menüs bei aktiver Auswahl).
Der Adapter nutzt nun durchgängig ThumbnailsCacheManager.getBitmapFromDiskCache(file) und executeThumbnailTask (210-305), setzt fehlende Thumbnails auf Platzhalter, räumt PNG-Hintergründe wieder frei und verhindert Starten der Generation ohne gültiges Konto.
Weitere Views (FileDetailsFragment.kt:430-444, MainFileListFragment.kt:603-620, RemoveFilesDialogFragment.kt:120-134, ShareFileFragment.kt:240, ReceiveExternalFilesAdapter.java:151) migrieren auf die neue Cache-API, sodass alle Thumbnail-Einstiegsstellen dieselbe Speicherstrategie und das zentrale Task-Dispatching verwenden.

zerox80 commented:
fix(auth): OIDC-Login ohne Dynamic Registration ermöglichen

- OAuthUtils.getClientAuth (opencloudApp/.../OAuthUtils.kt:87) gibt bei leerem Secret jetzt `null` zurück, damit kein leerer Basic-Header gesendet wird.
- LoginActivity.exchangeAuthorizationCodeForTokens (opencloudApp/.../LoginActivity.kt:605) wurde so angepasst, dass:
  - zuerst dynamisch registrierte Clients genutzt und andernfalls auf `oauth2_client_id` zurückgegriffen wird,
  - `client_id` immer im Token-Request-Body bleibt, auch bei Public Clients,
  - client_secret/Authorization nur angehängt werden, wenn ein nicht-leeres Secret vorliegt und der Server client_secret_post explizit unterstützt.
- AccountAuthenticator.performTokenRefresh (opencloudApp/.../AccountAuthenticator.java:344) übernimmt die gleiche Fallback-Logik beim Refresh Token und respektiert token_endpoint_auth_methods_supported.
- TokenRequest und TokenRequestParams (opencloudDomain/.../TokenRequest.kt:33, opencloudComLibrary/.../TokenRequestParams.kt:31) akzeptieren nun nullbare Auth-Header; TokenRequestRemoteOperation (opencloudComLibrary/.../TokenRequestRemoteOperation.kt:55) verzichtet entsprechend auf den Authorization-Header, wenn kein Secret vorhanden ist.
- Abgleich mit dem Backend: Der mitgelieferte Keycloak-Client (`opencloud/devtools/deployments/opencloud_full/config/keycloak/clients/OpenCloudAndroid.json`) ist als `publicClient: true` hinterlegt und passt somit zur neuen Fallback-Strategie.
- `gradlew`-Unit-Tests scheitern weiterhin an einer veralteten TUS-Integrationstest-Signatur; die Korrektur steht noch aus und hängt nicht mit dieser Änderung zusammen.

Refs: https://github.com/opencloud-eu/android/issues/48

ankumm commented:
> > Could you provid an apk for tester who are not able to compile it?
> 
> https://github.com/zerox80/android/releases/tag/1.0rc

Danke! Läuft wie geschmiert jetzt ;-)

zerox80 commented:
Wait, why from August? 

[zerox80](https://github.com/zerox80) commented [last week](https://github.com/opencloud-eu/android/pull/36#issuecomment-3441126881)
feat(upload): honor TUS capabilities and refine fallback logic

Read filesTusSupport from capabilities and pass into TUS flow
Respect server maxChunkSize and HTTP method override during PATCH
Prefer TUS for large files; on failure, force chunked fallback for
large files when TUS is advertised, even if legacy chunking is off
Expanded logging for chosen chunk sizes and overrides
Applied to ContentUri and FileSystem upload workers
Why: improve interoperability with varying servers/proxies, respect
server limits, and increase reliability of large uploads.

This solves lot of issues, for example using 5G Network etc. Ill keep this one aswell.

zerox80 commented:
@guruz

guruz commented on opencloudApp/src/main/java/eu/opencloud/android/workers/UploadFileFromContentUriWorker.kt:
```diff
@@ -252,14 +259,129 @@ class UploadFileFromContentUriWorker(
+        }
+
+        if (!usedTus) {
+            if (isChunkingAllowed && fileSize > CHUNK_SIZE) {
```
@TheOneRing says that actually the old chunking is being phased out. So basicaly, files are uploaded either with a single PUT or with TUS. So you can kick out this handling.

guruz commented on opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/CreateTusUploadRemoteOperation.kt:
```diff
@@ -0,0 +1,199 @@
+                    val parentDir = remotePath.substringBeforeLast('/', "/")
+                    buildString {
+                        append(client.baseUri.toString().trimEnd('/'))
+                        append("/remote.php/webdav")
```
This hardcoded string is fishy.

guruz commented on opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/CheckTusSupportRemoteOperation.kt:
```diff
@@ -0,0 +1,93 @@
+                    }
+
+                    // Fallback probe on this endpoint: try POST with minimal test
+                    Timber.d("TUS headers inconclusive at %s, probing with POST test", endpoint)
```
I don't think a TUS probe is necessary. OpenCloud always supports TUS(?).

guruz commented on opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/CreateTusUploadRemoteOperation.kt:
```diff
@@ -0,0 +1,199 @@
+            // TUS Upload-Metadata (decoded for debugging)
+            val filename = remotePath.substringAfterLast('/')
+            val mtime = (file.lastModified() / 1000).toString()
+            val checksum = "sha1 " + computeSha1(file)
```
Why is there a sha1 computed here?
Is the checksum (sha256) not coming from the caller of this function?
Maybe I'm misunderstanding something

guruz reviewed (changes requested):
Hi, my comments are added to the individual lines. 

**I have only looked at the first commit dd872cd5fe610d2431fcc7b3583b0b2788f0cf53 so far and not at your other stuff .**

Regarding what you mean about Progress display, I think the function _getRunningUploadsWorkInfosLiveData_ might be relevant..

zerox80 commented on opencloudComLibrary/src/main/java/eu/opencloud/android/lib/resources/files/tus/CheckTusSupportRemoteOperation.kt:
```diff
@@ -0,0 +1,93 @@
+                val idx = initial.indexOf("/uploads/")
+                if (idx > 0) initial.substring(0, idx + "/uploads/".length) else withSlash
+            }
+            val altInitial = initial.replace("/remote.php/dav/", "/dav/")
```
Updated this as well. The OPTIONS request now just hits the uploads collection (base + optional trailing slash) and drops the manual `/remote.php` replacements. That keeps the detection logic in line with what the backend serves and avoids the brittle string juggling.

zerox80 commented on opencloudApp/src/main/java/eu/opencloud/android/workers/UploadFileFromFileSystemWorker.kt:
```diff
@@ -126,6 +132,161 @@ class UploadFileFromFileSystemWorker(
+
+                    // Try to recover by re-checking current offset from server and continue
+                    val recover = try {
+                        val off = GetTusUploadOffsetRemoteOperation(tusUrl!!).execute(client)
```
Yep, I routed the head/creation/recover calls through executeRemoteOperation inside TusUploadHelper, so we still get the mapped exceptions while keeping the loop logic intact.

zerox80 commented:
The old WebDAV chunking logic has been removed – we now only try TUS (if the file size is sufficient) and otherwise fall back to a simple PUT.
The TUS-specific upload logic is now located in TusUploadHelper, so that UploadFileFromFileSystemWorker and UploadFileFromContentUriWorker use the same path.
CheckTusSupportRemoteOperation works without hard-coded /remote.php replacements; the OPTIONS request goes directly to the upload collection, fallback PROBE removed.
The checksum now only uses the value provided by the caller; the internal SHA-1 code has been removed.
I deleted the local tusUploadOffset – when continuing, we query the server via GetTusUploadOffsetRemoteOperation.
All TUS requests run via executeRemoteOperation so that errors are mapped consistently as before.
The 404 in PATCH came from the incorrectly assembled upload path; the last commit also corrects this.

guruz commented:
> Please review,

Will definitely do :-) Today or tomorrow... thanks for the work

zerox80 commented on opencloudApp/src/main/java/eu/opencloud/android/presentation/files/filelist/FileListAdapter.kt:
```diff
@@ -85,13 +91,22 @@ class FileListAdapter(
+            is OCFileWithSyncInfo -> item.file.id ?: item.file.remotePath.hashCode().toLong()
+            is OCFooterFile -> Long.MIN_VALUE + position
+            else -> position.toLong()
+        }
```
Wir hatten setHasStableIds(true) schon aktiv, aber ohne echte IDs. Die Implementierung hier bewahrt Selektionszustände & verhindert flackernde Updates, wenn DiffUtil läuft.

zerox80 commented on opencloudApp/src/main/java/eu/opencloud/android/presentation/files/filelist/FileListAdapter.kt:
```diff
@@ -109,7 +124,7 @@ class FileListAdapter(
                     tag = ViewType.GRID_ITEM
-                    filterTouchesWhenObscured = PreferenceUtils.shouldDisallowTouchesWithOtherVisibleWindows(context)
+                    filterTouchesWhenObscured = disallowTouchesWithOtherWindows
                 }
```
disallowTouchesWithOtherWindows hole ich jetzt einmal und setze es in jedem Holder. Vorher wurde die Preference mehrfach abgefragt und List/Grid hatten leicht unterschiedliches Verhalten

zerox80 commented on opencloudApp/src/main/java/eu/opencloud/android/presentation/files/filelist/FileListAdapter.kt:
```diff
@@ -166,33 +183,44 @@ class FileListAdapter(
 
                 // Allow or disallow touches with other visible windows
-                filterTouchesWhenObscured = PreferenceUtils.shouldDisallowTouchesWithOtherVisibleWindows(context)
+                filterTouchesWhenObscured = disallowTouchesWithOtherWindows
```
Cache den Overlay-Flag jetzt einmal (disallowTouchesWithOtherWindows) und nutze ihn für alle Unterviews; gleichzeitig hole ich das Thumbnail wieder direkt über getBitmapFromDiskCache(file) – der remoteId-Umweg war unnötig

guruz commented on opencloudApp/src/main/java/eu/opencloud/android/workers/TusUploadHelper.kt:
```diff
@@ -0,0 +1,272 @@
+            )
+            lastModified?.takeIf { it.isNotBlank() }?.let { metadata["mtime"] = it }
+
+            val sha256 = runCatching { computeSha256Hex(localPath) }
```
Here again the sha256 is computed via FileInputStream even though the caller supplies it in transfer.tusUploadChecksum?

zerox80 replied:
ur right, its solved now

guruz commented on opencloudData/src/main/java/eu/opencloud/android/data/OpencloudDatabase.kt:
```diff
@@ -50,6 +50,7 @@ import eu.opencloud.android.data.migrations.MIGRATION_35_36
 import eu.opencloud.android.data.migrations.MIGRATION_37_38
 import eu.opencloud.android.data.migrations.MIGRATION_41_42
 import eu.opencloud.android.data.migrations.MIGRATION_42_43
+import eu.opencloud.android.data.migrations.MIGRATION_48_49
```
Why is there a 48.json and a 49.json?
For a single commit or PR, shouldn't it be one new version?

zerox80 commented:
11-05 18:32:44.594 D/TusUploadHelper(30609): TUS: starting upload for /100MB (1).bin size=104857600
11-05 18:32:44.595 D/TusUploadHelper(30609): TUS: creating upload resource filename=100MB (1).bin size=104857600 metadata={filename=100MB (1).bin, mimetype=bin}
11-05 18:32:44.596 D/TusUploadHelper(30609): TUS: using collection endpoint: https://oc.hidden.eu/dav/spaces/0e465bba-14f2-4786-9917-7f51bd25a5d9$fbc2142d-77eb-4f92-b3c9-5be2bb277445
11-05 18:32:44.598 D/CreateTusUploadRemoteOperation(30609): TUS resolved collection: https://oc.hidden.eu/dav/spaces/0e465bba-14f2-4786-9917-7f51bd25a5d9$fbc2142d-77eb-4f92-b3c9-5be2bb277445
11-05 18:32:44.599 D/CreateTusUploadRemoteOperation(30609): TUS Creation URL: https://oc.hidden.eu/dav/spaces/0e465bba-14f2-4786-9917-7f51bd25a5d9$fbc2142d-77eb-4f92-b3c9-5be2bb277445
11-05 18:32:47.050 D/CreateTusUploadRemoteOperation(30609): TUS Creation [https://oc.hidden.eu/dav/spaces/0e465bba-14f2-4786-9917-7f51bd25a5d9$fbc2142d-77eb-4f92-b3c9-5be2bb277445] - 201
11-05 18:32:47.054 D/CreateTusUploadRemoteOperation(30609): TUS upload resource created: https://oc.hidden.eu/data/<token redacted>
11-05 18:32:47.453 D/TusUploadHelper(30609): TUS: resume offset 10485760 / 104857600
11-05 18:32:47.458 D/TusUploadHelper(30609): TUS: uploading chunk=10000000 remaining=94371840
11-05 18:33:02.861 D/TusUploadHelper(30609): TUS: uploading chunk=10000000 remaining=84371840

It works

zerox80 commented on opencloudApp/src/main/java/eu/opencloud/android/workers/TusUploadHelper.kt:
```diff
@@ -200,26 +202,17 @@ class TusUploadHelper(
+        
+        Timber.d("TUS: using collection endpoint: %s", normalizedBase)
+        
+        return normalizedBase
```
Ich habe mir die Android-TUS-Logs genauer angeschaut. Dabei ist mir aufgefallen, dass die App beim HTTP‑POST zum Erstellen des TUS-Uploads den normalen WebDAV-Pfad (/remote.php/dav/files/...) beziehungsweise den Space-Pfad mit abschließendem Slash benutzt hat. Genau diese URL versteht der Server aber nicht für TUS – er erwartet den Space-Endpunkt ohne Slash.

Sobald ich das auf den korrekten Endpunkt (/dav/spaces/<spaceId>) ohne Slash umgestellt habe, kam direkt die 201‑Antwort. Zusammen mit dem ersten Chunk (creation-with-upload) lief der Upload dann sauber durch.

Sprich: Der eigentliche Fehler war, dass wir den falschen WebDAV-Pfad fürs TUS-Create genommen haben.

zerox80 commented on opencloudApp/src/main/java/eu/opencloud/android/workers/TusUploadHelper.kt:
```diff
@@ -68,14 +68,16 @@ class TusUploadHelper(
-                    useCreationWithUpload = false,
-                    firstChunkSize = null,
+                    useCreationWithUpload = true,
+                    firstChunkSize = firstChunkSize,
```
Hier schalte ich creation-with-upload aktiv – genauso macht es der Web-Client. Damit schicken wir direkt die ersten 10 MB im POST mit, der Server antwortet zuverlässig mit 201 und wir sparen uns den extra Request ohne Payload, der zuvor 404/500 geliefert hat.

zerox80 commented:
> @guruz Please review every commit, before u ask for changes, maybe they're solved already in the next commits! Is that fine? just make ur notes and review the next commit, cuz i made lot of changes. I fixed some bugs, race conditioning, wrong tus upload path etc. But now its final, i guess. I fixed the issues u told me

11-06 08:06:10.063 D/TusUploadHelper( 4671): TUS: uploading chunk=10000000 at offset=1030485760 remaining=43256064
11-06 08:06:11.846 D/TusUploadHelper( 4671): TUS: uploading chunk=10000000 at offset=1040485760 remaining=33256064
11-06 08:06:13.700 D/TusUploadHelper( 4671): TUS: uploading chunk=10000000 at offset=1050485760 remaining=23256064
11-06 08:06:15.593 D/TusUploadHelper( 4671): TUS: uploading chunk=10000000 at offset=1060485760 remaining=13256064
11-06 08:06:17.405 D/TusUploadHelper( 4671): TUS: uploading chunk=3256064 at offset=1070485760 remaining=3256064
11-06 08:06:22.198 I/TusUploadHelper( 4671): TUS: upload completed for /test/1GB.bin (size=1073741824)

now without bugs.

kulmann commented:
Really happy about this PR existing, but still, I need to intervene here (sorry 🥲 ). This PR is far too large (both lines of code and complexity wise) to give it a decent review. @zerox80 could you try to either come up with a new PR with the smallest change set possible, or remove anything from this PR goes beyond the smallest change set possible? At least the thumbnail code really doesn't belong in this PR and maybe the foreground service support for upload workers should also be a followup. Hope this doesn't demotivate you, but I don't see this PR ever getting merged with the current scope. Thanks a lot!

zerox80 commented:
> Really happy about this PR existing, but still, I need to intervene here (sorry 🥲 ). This PR is far too large (both lines of code and complexity wise) to give it a decent review. @zerox80 could you try to either come up with a new PR with the smallest change set possible, or remove anything from this PR goes beyond the smallest change set possible? At least the thumbnail code really doesn't belong in this PR and maybe the foreground service support for upload workers should also be a followup. Hope this doesn't demotivate you, but I don't see this PR ever getting merged with the current scope. Thanks a lot!

Hi @kulmann,

I've reduced the PR scope as you suggested! 

## What I did:
- Removed 6 commits from this PR
- Now contains only **2 commits** with the core TUS upload implementation
- Reduced from ~3500 to ~2870 lines of changes
- Split removed features into separate PRs

## Commits in this PR now:
1. `feat(upload): implement TUS resumable upload protocol` - Core TUS 1.0.0 implementation
2. `chore: update Gradle build configuration settings` - Required gradle setting

## Split features into separate PRs:
- Thumbnail cache improvements → PR #50
-  Upload cancellation support → PR #53 (depends on this PR)
-  Foreground service & retry improvements → PR #54 (depends on this PR)

## What's in this PR:
- TUS 1.0.0 protocol implementation
- Chunked uploads with resume capability
- SHA-256 checksum verification
- Database migration (v49) for TUS state persistence
- TUS capability detection
- Fallback to standard PUT if TUS unavailable
- Comprehensive integration tests

## Note:
PR #50 (thumbnail cache) is completely independent and can be merged immediately.
PRs #53 and #54 depend on this PR and will be rebased after this is merged.

Thanks for the feedback! 😊

zerox80 commented:
> Thanks! i'm still in general review before I can test. One question: What about a use case where a source file suddenly changes size? Will this invalidate the TUS upload? Or is this not relevant for the current upload architecture?

If the source file changes size during upload, it will cause issues because TUS relies on the fixed Upload-Length declared at the start.

If file shrinks: The upload will fail. The client will try to read past the new EOF to fulfill the promised length, causing an IO/Protocol exception.
If file grows: The upload will succeed but be truncated. The client stops at the original Upload-Length, ignoring the new appended data.
So yes, changing the source file invalidates the consistency of the upload. Ideally, we should upload a snapshot or ensure the file is locked/immutable during the process.

guruz commented:
> If the source file changes size during upload, it will cause issues because TUS relies on the fixed Upload-Length declared at the start.

And also the server should fail because of the sha256 checksum mismatch, right @TheOneRing ?

guruz commented:
... or actually maybe i'm confused now about checksum for the whole file.
vs
checksum for the single chunk?

guruz commented:
@zerox80 Can you explain where are you sending the checksum to the server when creating the TUS upload?
I see this here but I feel this is too late: https://github.com/opencloud-eu/android/pull/36/commits/3dbfdc761e484c871a65d0b76fa37929ffbcb8b7#diff-74a99898803c7adb3b503022c396d7030acfc9e5d83c173892fc74648ae8b8d3R98

Did you check your request log output if you are really sending it?

I can see the desktop client sends it on creation https://github.com/opencloud-eu/desktop/blob/90aea3a23882e05dab82d8b72cedfcd1425fb366/src/libsync/propagateuploadtus.cpp#L85

zerox80 commented:
Regarding the file headers: I explicitly grant permission to use 'OpenCloud.eu' as the author in the new files instead of my personal GitHub username. I do not require personal attribution, so please feel free to standardize the headers as you see fit.
@guruz

zerox80 commented:
I hereby assign the copyright attribution for these changes to OpenCloud GmbH. You are welcome to list 'OpenCloud GmbH' in the file headers/license texts instead of my name.

zerox80 commented:
Hi, u can merge this, and we can discuss later about this, is this fine?
I tried to make it similar to the web. Thefirst Request is a  Post Request - a10MB Cunk, like the web does.
Or do u mean something else?

guruz commented:
> but as you suggested, let's do that in a follow-up PR to keep this one focused. 

Yes 100% please keep this PR as it is.
I will merge it today/tomorrow. Sorry this all takes longer than expected ... **_thank you for your contribution!_**
